bundler/lib/bundler/digest.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+# This code was extracted from https://github.com/Solistra/ruby-digest which is under public domain
+module Bundler
+  module Digest
+    # The initial constant values for the 32-bit constant words A, B, C, D, and
+    # E, respectively.
+    SHA1_WORDS = [0x67452301, 0xEFCDAB89, 0x98BADCFE, 0x10325476, 0xC3D2E1F0].freeze
+
+    # The 8-bit field used for bitwise `AND` masking. Defaults to `0xFFFFFFFF`.
+    SHA1_MASK = 0xFFFFFFFF
+
+    class << self
+      def sha1(string)
+        unless string.is_a?(String)
+          raise TypeError, "can't convert #{string.class.inspect} into String"
+        end
+
+        buffer = string.b
+
+        words = SHA1_WORDS.dup
+        generate_split_buffer(buffer) do |chunk|
+          w = []
+          chunk.each_slice(4) do |a, b, c, d|
+            w << (((a << 8 | b) << 8 | c) << 8 | d)
+          end
+          a, b, c, d, e = *words
+          (16..79).each do |i|
+            w[i] = SHA1_MASK & rotate((w[i-3] ^ w[i-8] ^ w[i-14] ^ w[i-16]), 1)
+          end
+          0.upto(79) do |i|
+            case i
+            when  0..19
+              f = ((b & c) | (~b & d))
+              k = 0x5A827999
+            when 20..39
+              f = (b ^ c ^ d)
+              k = 0x6ED9EBA1
+            when 40..59
+              f = ((b & c) | (b & d) | (c & d))
+              k = 0x8F1BBCDC
+            when 60..79
+              f = (b ^ c ^ d)
+              k = 0xCA62C1D6
+            end
+            t = SHA1_MASK & (SHA1_MASK & rotate(a, 5) + f + e + k + w[i])
```

Comment:
As `+` is higher precedence than `&`, this equals `SHA1_MASK & (SHA1_MASK & (rotate(a, 5) + f + e + k + w[I]))`.
I don't think bit-and with the same value twice has meaning.
Maybe `SHA1_MASK & ((SHA1_MASK & rotate(a, 5)) + f + e + k + w[I])` is intended?

Reply:
[rolling own crypto is hard](https://github.com/rubygems/rubygems/pull/4989#issuecomment-944224157) :) Anyway isn't this used mostly for caching? There could be problem, but not visible to clients (except collision case).

Reply:
Yes, sure, what I'm saying is that it seems better to not change the behaviour without fully understanding the bug. If @nobu can provide a string incorrectly digested with the current code, then of course we should fix it.

Reply:
We have a test suite that ensure this vendored version produce the same hashes than the stdlib digest. If we had a bug here we'd have widely different output.

It's safe to assume there's no bug, just a small ineficiency.

Reply:
@byroot this is the CI covering this.

```ruby
["foo", "skfjsdlkfjsdf", "3924m", "ldskfj"].each do |payload|
  expect(subject.sha1(payload)).to be == stdlib.hexdigest(payload)
end
```

In theory it is still possible some other strings could be hashed wrongly. Indeed if @nobu can provide one, we need to fix this.

---

As you can see rolling own crypto and maintaining it is hard. I still think merging this was not good idea (even it is great work @byroot :heart:, I'm not complaining about the code itself).

Reply:
The above test is enough to cover that code, hash functions have widely different outputs.

Also it's not hard to understand that `a & a == a`. Let's not be dogmatic here.

> I still think merging this was not good idea 

What else would have you done? `digest` was gemified, by design `bundler` can't rely on gems before the full gemfile is activated.

We talked about using something else than SHA1 for this, and IIRC it's not even for crypto but just mangling, but we ruled it out as it would have broken backward compatibility.

Reply:
We needed a solution for a problem, @byroot came up with this one, and this works well. I don't see the issue here, to. be honest. Let's add this simplification and move on, no?

Reply:
https://github.com/rubygems/rubygems/pull/5595

Reply:
> Let's add this simplification and move on, no?

Sure, I'll check if there is any way to move out of MD5 (and digest at all) and we can start slowly deprecate that somehow. :pray: